### DISCUSSION TITLE
Access updated state when handling events

### DISCUSSION THREAD

**Discussion by cwsiteplan** [cashapp/molecule#471] [Q&A] [answered]
Problem when i want to read state in the event handling.
As the nature of LaunchedEffect it captures the values upon "launch" the values read are empty

```
@Composable
fun pupperPicsPresenter(events: Flow<Event>, service: PupperPicsService): Model {
  var breeds: List<String> by remember { mutableStateOf(emptyList()) }

  // Grab the list of breeds and sets the current selection to the first in the list.
  // Errors are ignored in this sample.
  LaunchedEffect(Unit) {
    breeds = service.listBreeds()
  }

  // Handle UI events.
  LaunchedEffect(Unit) {
    events.collect { event ->
      when (event) {
        is Event.ChangeBreed -> breeds.first { ... }  // <-- breeds is empty as the launched effect captures initial value
        is Event.AdjustNewBreed -> ...
      }
    }
  }
...
}
```

if I change the key for the launched effect to be updated when the key updates like this:

```
@Composable
fun pupperPicsPresenter(events: Flow<Event>, service: PupperPicsService): Model {
  var breeds: List<String> by remember { mutableStateOf(emptyList()) }

  // Grab the list of breeds and sets the current selection to the first in the list.
  // Errors are ignored in this sample.
  LaunchedEffect(Unit) {
    breeds = service.listBreeds()
  }

  // Handle UI events.
  LaunchedEffect(breeds) {
    events.collect { event ->
      when (event) {
        is Event.ChangeBreed -> breeds.first { ... }  // <-- breeds are fine now
        is Event.AdjustNewBreed -> ... // <- event is dropped as launched effect got re-initialized
      }
    }
  }
...
}
```

i do miss other events that are triggered once a new item is added.

Has anyone else dealt with that situation?

**Comment by JakeWharton** [comment, marked as answer]
I believe this is what `rememberUpdatedState` is meant to solve: https://developer.android.com/develop/ui/compose/side-effects#rememberupdatedstate

**Comment by cwsiteplan** [comment]
oh, thanks - wasn't aware of that - works fine! 👌  🙏

**Reply by mwilliams-sfo** [reply]
@cwsiteplan I have a similar problem and it would be helpful if you could describe how you used rememberUpdatedState to solve the above problem. Since you modify the state variable, and rememberUpdatedState returns read-only state, I don't see the connection.

**Reply by cwsiteplan** [reply]
i have a dedicated property for holding the updated readable state.
i do not "write" this property - it's just holding the updated values (from DB)
```
val annotations: List<MapAnnotation> = annotationFlow.collectAsState(emptyList()).value
val updatedAnnotations = rememberUpdatedState(annotations)
```
but i guess you can just wrap any mutable state in the same way.
